lib/improver/blending/weighted_blend.py

```diff
@@ -810,25 +688,113 @@ def weighted_mean(self, cube, weights):
+        else:
+            frt_calendar = frt_coord.units.calendar
+            cycletime_point = cycletime_to_number(
+                    cycletime, time_unit=frt_units, calendar=frt_calendar)
+            return np.round(cycletime_point).astype(np.int64)
+
+    def _set_coords_to_remove(self, input_cube):
+        """
+        Generate a list of coordinate names associated with the blend
+        dimension.  Unless these are time-related coordinates, they should be
+        removed after blending.
+
+        Args:
+            input_cube (iris.cube.Cube):
+                Cube to be blended
+        """
+        time_coords = ["time", "forecast_reference_time", "forecast_period"]
+        blend_dim, = input_cube.coord_dims(self.blend_coord)
+        self.crds_to_remove = []
+        for coord in input_cube.coords():
+            if coord.name() in time_coords:
+                continue
+            if blend_dim in input_cube.coord_dims(coord):
+                self.crds_to_remove.append(coord.name())
+
+    def _set_forecast_reference_time_and_period(self, blended_cube):
+        """
+        For cycle and model blending, update the forecast reference time and
+        forecast period coordinate points to the single most recent value,
+        rather than the blended average, and remove any bounds from the
+        forecast reference time. Modifies cube in place.
+
+        Args:
+            blended_cube (iris.cube.Cube)
+        """
+        blended_cube.coord("forecast_reference_time").points = [
+            self.cycletime_point]
+        blended_cube.coord("forecast_reference_time").bounds = None
+        if blended_cube.coords("forecast_period"):
+            blended_cube.remove_coord("forecast_period")
+        new_forecast_period = forecast_period_coord(blended_cube)
+        time_dim = blended_cube.coord_dims("time")
+        blended_cube.add_aux_coord(new_forecast_period, data_dims=time_dim)
+
+    def _update_blended_metadata(self, blended_cube, attributes_dict):
+        """
+        Update metadata after blending:
+        - For cycle and model blending, set a single forecast reference time
+        and period using self.cycletime_point or the latest cube contributing
+        to the blend
+        - Remove scalar coordinates that were previously associated with the
+        blend dimension
+        - Update attributes as specified via process arguments
+        Modifies cube in place.
+
+        Args:
+            blended_cube (iris.cube.Cube)
+            attributes_dict (dict or None)
+        """
+        if self.blend_coord in ["forecast_reference_time", "model_id"]:
+            self._set_forecast_reference_time_and_period(blended_cube)
+        for coord in self.crds_to_remove:
+            blended_cube.remove_coord(coord)
+        if attributes_dict is not None:
+            amend_attributes(blended_cube, attributes_dict)
+
+    def process(self, cube, weights=None,
+                cycletime=None, attributes_dict=None):
```

Comment:
There's no written guidance on this for IMPROVER and I think we all do this differently.  My logic for what I pass into `__init__` vs `process` is dividing the plugin between what it **is** and what it **does**.  So in this case:

- I initialise a plugin that will blend something over `blend_coord`
- I process a cube that wants to have a given cycle time and attributes
- I could use the same plugin to process a different cube, with a different cycle time and output attributes, if I wanted to

This is a bit convoluted as I'm struggling to put the concept into words - but I suppose another way of saying it would be that both `cycletime` and `attributes_dict` are about the specific cube that is being processed, and would probably be different for any other input cube; but the plugin is an instance that blends over `coord` and that's independent of the specific cube you pass in.

Does that make sense?

Reply:
Thanks for the explanation. Typically, I would just make the distinction cube vs non-cube when deciding what to put in the `__init__` vs the `process`, as I'm not typically setting up an instance and then calling it repeatedly, which is when the logic that you describe would be more relevant. We could possibly chat about this as a group at some point.

Reply:
You're right, that's not how we use the plugins at the moment.  But where there's no scientific reason not to run a "process" more than once, I would always prefer to design plugins so that they can operate several times if need be.  Then for example someone could pass all their diagnostics into the same blending "CLI" / suite step, and have the same plugin instance process all of those diagnostics.  It's not our current design, but why would we write our code to be closed to this sort of usage?

Reply:
And yes, this is probably something we should talk through as a group.  I'll raise it with @arh89  for a developer meeting.

Reply:
> My logic for what I pass into `__init__` vs `process` is dividing the plugin between what it **is** and what it **does**. 
I like this idea - although I recognise that it's a bit "fluffy"/hard to describe.

> Typically, I would just make the distinction cube vs non-cube when deciding what to put in the `__init__` vs the `process`, as I'm not typically setting up an instance and then calling it repeatedly, which is when the logic that you describe would be more relevant.
I think a lot of people have gone for cube vs non-cube, which is perfectly valid, but I do like what @cgsandford suggested above. To be honest, it's been a long-standing subset-of-Green-team (me, @LaurenceBeard and @TomekTrzeciak) grumble that we create these plugins and only use them once - which isn't harmful, because they take no time at all to set up, but it does affect how we think about the separation between `__init__` and `process`. You'll find that we sit on different sides of the fence though - @TomekTrzeciak is quite anti-plugin class (he'd rather we just have a function), whereas I'm quite pro it (provided that we make use of them sensibly).

With respect to my personal view of how I would like my ideal IMPROVER architecture (which may disagree with @TomekTrzeciak and @benfitzpatrick), I'd like that we did create these plugins up-front, which is what @cgsandford is suggesting, and then use them as callables in the compute graph. That's one of the reasons why I asked @neilCrosswaite to add `__call__` to his `BasePlugin` class (now merged). Essentially, creating the plugin and setting up the options should (imo) be roughly equivalent to `partial`-ing a function.

> And yes, this is probably something we should talk through as a group.
Yep, but I wouldn't expect to arrive at any consensus, and the conversation will likely end up very technical - just a heads up.